server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportUnpromotableShardRefreshAction.java

```diff
@@ -73,6 +76,17 @@ protected void unpromotableShardOperation(
             return;
         }
 
+        // During an upgrade to FAST_REFRESH_RCO, we expect search shards to be first upgraded before the primary is upgraded. Thus,
+        // when the primary is upgraded, and starts to deliver unpromotable refreshes, we expect the search shards to be upgraded already.
+        // TODO: remove assertion (ES-9563)
```

Comment:
I'd note that the fast refresh setting is final, since otherwise this could in theory fail.